core/store/src/trie/split_state.rs

```diff
@@ -67,68 +161,125 @@ impl ShardTries {
+fn apply_delayed_receipts_to_split_states_impl<'a>(
+    trie_updates: &mut HashMap<ShardUId, TrieUpdate>,
+    insert_receipts: &[Receipt],
+    delete_receipts: &[Receipt],
+    account_id_to_shard_id: &(dyn Fn(&AccountId) -> ShardUId + 'a),
+) -> Result<(), StorageError> {
+    let mut delayed_receipts_indices_by_shard = HashMap::new();
+    for (shard_uid, update) in trie_updates.iter() {
+        delayed_receipts_indices_by_shard.insert(*shard_uid, get_delayed_receipt_indices(update)?);
+    }
+
+    for receipt in insert_receipts {
+        let new_shard_uid: ShardUId = account_id_to_shard_id(&receipt.receiver_id);
+        if !trie_updates.contains_key(&new_shard_uid) {
+            let err = format!(
+                "Account {} is in new shard {:?} but state_roots only contains {:?}",
+                receipt.receiver_id,
+                new_shard_uid,
+                trie_updates.keys(),
             );
-            delayed_receipts_indices.next_available_index =
-                delayed_receipts_indices.next_available_index.checked_add(1).ok_or_else(|| {
-                    StorageError::StorageInconsistentState(
-                        "Next available index for delayed receipt exceeded the integer limit"
-                            .to_string(),
-                    )
-                })?;
+            return Err(StorageError::StorageInconsistentState(err));
         }
+        // we already checked that new_shard_uid is in trie_updates and delayed_receipts_indices
+        // so we can safely unwrap here
+        let mut delayed_receipts_indices =
+            delayed_receipts_indices_by_shard.get_mut(&new_shard_uid).unwrap();
+        set(
+            trie_updates.get_mut(&new_shard_uid).unwrap(),
+            TrieKey::DelayedReceipt { index: delayed_receipts_indices.next_available_index },
+            receipt,
+        );
+        delayed_receipts_indices.next_available_index =
+            delayed_receipts_indices.next_available_index.checked_add(1).ok_or_else(|| {
+                StorageError::StorageInconsistentState(
+                    "Next available index for delayed receipt exceeded the integer limit"
+                        .to_string(),
+                )
+            })?;
+    }
 
-        // commit the trie_updates and update state_roots
-        let mut merged_store_update = StoreUpdate::new_with_tries(self.clone());
-        let mut new_state_roots = HashMap::new();
-        for (shard_uid, mut trie_update) in trie_updates {
-            set(
-                &mut trie_update,
-                TrieKey::DelayedReceiptIndices,
-                delayed_receipts_indices_by_shard.get(&shard_uid).unwrap(),
+    for receipt in delete_receipts {
+        let new_shard_uid: ShardUId = account_id_to_shard_id(&receipt.receiver_id);
+        if !trie_updates.contains_key(&new_shard_uid) {
+            let err = format!(
+                "Account {} is in new shard {:?} but state_roots only contains {:?}",
```

Comment:
```suggestion
                "Account {:?} is in new shard {:?} but state_roots only contains {:?}",
```

It's interesting that package compiles successfully but CLion claims that `AccountId` doesn't implement Display. The same applies to line 223 above.

Maybe cc @matklad

Reply:
hmm I'm not sure why, it works in my CLion

Reply:
Oh, I had to update my version of CLion and then it worked.